### ISSUE TITLE
UnhandledPromiseRejectionWarning when not able to connect to redis

### ISSUE THREAD

**Issue by rajat12a** [taskforcesh/bullmq#521]
BullMQ Version: 1.24.0
Node Version: v14.16.0
NPM Version: 7.10.0

Summary: 
When redis is not able to connect to the redis host machine then it is throwing UnhandledPromiseRejectionWarning along with unhandled ioredis error. 

Steps to reproduce:

```
// test.js
const { QueueScheduler } = require('bullmq')

const testQueueScheduler = new QueueScheduler('test', {
  connection: {
    host: '1.0.0.0',
    port: 6379,
    db: 1
  }
})

testQueueScheduler.on('error', function (error) {
  console.error('caught the error', error)
})
```

What to expect:
The ETIMEDOUT error to be handled by on error function something like this
```
caught the error Error: connect ETIMEDOUT
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/index.js:308:37)
    at Object.onceWrapper (events.js:421:28)
    at Socket.emit (events.js:315:20)
    at Socket._onTimeout (net.js:483:8)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7) {
  errorno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  syscall: 'connect'
}
```

What is happening:
```
caught the error Error: connect ETIMEDOUT
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/index.js:308:37)
    at Object.onceWrapper (events.js:421:28)
    at Socket.emit (events.js:315:20)
    at Socket._onTimeout (net.js:483:8)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7) {
  errorno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  syscall: 'connect'
}
(node:53555) UnhandledPromiseRejectionWarning: Error: connect ETIMEDOUT
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/index.js:308:37)
    at Object.onceWrapper (events.js:421:28)
    at Socket.emit (events.js:315:20)
    at Socket._onTimeout (net.js:483:8)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:53555) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:53555) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[ioredis] Unhandled error event: Error: connect ETIMEDOUT
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/index.js:308:37)
    at Object.onceWrapper (events.js:421:28)
    at Socket.emit (events.js:315:20)
    at Socket._onTimeout (net.js:483:8)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)

```

There are unwanted error logs even though testQueueScheduler is being handled with on emit event, the same happens with Worker and Queue as well

Also on connecting to a redis-server and then bringing that redis-server down, so on reconnecting ETIMEDOUT error is being caught but then after certain retries this UnhandledPromiseRejectionWarning pops up
```
caught the error Error: connect ETIMEDOUT
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/index.js:308:37)
    at Object.onceWrapper (events.js:421:28)
    at Socket.emit (events.js:315:20)
    at Socket._onTimeout (net.js:483:8)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7) {
  errorno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  syscall: 'connect'
}
(node:57411) UnhandledPromiseRejectionWarning: MaxRetriesPerRequestError: Reached the max retries per request limit (which is 20). Refer to "maxRetriesPerRequest" option for details.
    at Socket.<anonymous> (/home/test/node_modules/ioredis/built/redis/event_handler.js:176:37)
    at Object.onceWrapper (events.js:422:26)
    at Socket.emit (events.js:315:20)
    at TCP.<anonymous> (net.js:673:12)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:57411) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:57411) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by aarhusgregersen** [comment]
I'm experiencing this exact issue on latest BullMQ, ioredis and node 18.11.0. 
I get an ETIMEDOUT in production, as if its unable to connect to redis, but redis otherwise working normally. The newer version of bullMQ does not use the queue-scheduler, but I'm still seeing the issue. 

In my case the minimal reproduction looks like this (I've filtered out other pieces of the code in each file):
./app.ts
```
import cors from "cors";
import dotenv from "dotenv";
import express from "express";
import session from "express-session";
import useragent from "express-useragent";
import morgan from "morgan";
import passport from "passport";
import path from "path";
const RedisStore = require("connect-redis")(session);
const helmet = require("helmet");

dotenv.config({ debug: true });
const app: express.Application = express();
const utils = require("./src/common/helpers/utils");
const auth = require("./src/common/auth/helpers/authHelper");

import { createBullBoard } from "@bull-board/api";
import { BullMQAdapter } from "@bull-board/api/bullMQAdapter";
import { ExpressAdapter } from "@bull-board/express";
import { corsSettings } from "./src/common/auth/helpers/corsSettings";
import whilelistDomains from "./src/common/auth/helpers/whitelist.json";
import { redisClient } from "./src/common/helpers/db";
import { jobRunner } from "./src/common/jobs/jobRunner";
... 

// Connect to redis
redisClient.on("connect", () => {
	console.log("Connected to redis");
});

const serverAdapter = new ExpressAdapter();
serverAdapter.setBasePath("/admin/queues");

createBullBoard({
	queues: [
		new BullMQAdapter(networkUnreadsQueue),
		new BullMQAdapter(taskDigestsQueue),
		new BullMQAdapter(remindersQueue),
		new BullMQAdapter(userActivationRemindersQueue),
		new BullMQAdapter(sendMessagesQueue),
	],
	serverAdapter: serverAdapter,
});

app.use("/admin/queues", serverAdapter.getRouter());

jobRunner();

// Get port from environment and store in Express.
const port = normalizePort(process.env.PORT);
app.set("port", port);

// Listen on provided port, on all network interfaces.
const server = app.listen(port, function () {
	log.info(
		`Listening on ${process.env.ENV_URL_API} on ${process.env.NODE_ENV}`,
	);
});
...
```
jobrunner.ts: 
```
import { scheduledAssignmentsQueue } from "./queues/scheduledAssignments.queue";
import { userActivationRemindersQueue } from "./queues/userActivationReminders.queue";
import { remindersQueue } from "./queues/reminders.queue";
import { newTimelineSectionsQueue } from "./queues/newTimelineSection.queue";
import { networkUnreadsQueue } from "./queues/networkUnreads.queue";
import { taskRemindersQueue } from "./queues/taskReminders.queue";
import { sendMessagesQueue } from "./queues/sendMessage.queue";

const env = process.env.NODE_ENV || "development";

export const jobRunner = async () => {
	try {
		if (env === "production") {
			console.log("[JOBS] Jobs being scheduled in production...");

			await scheduledAssignmentsQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "0 2 * * 1-5" },
				},
			);
			await userActivationRemindersQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "30 2 * * 1-5" },
				},
			);
			await remindersQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "0 3 * * 1-5" },
				},
			);
			await newTimelineSectionsQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "30 3 * * 1-5" },
				},
			);
			await networkUnreadsQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "0 4 * * 1-5" },
				},
			);
			await taskRemindersQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "30 4 * * 1-5" },
				},
			);
			await sendMessagesQueue.add(
				"Scheduled Job",
				{},
				{
					repeat: { pattern: "30 7 * * 1-5" },
				},
			);
                  }
	} catch (err) {
		console.log(`${env.toUpperCase}: Could not start job runner.`);
		console.log(err);
	}
};
```

Here's an example queue, userActivationReminders:
```
import { Job, Queue, Worker } from "bullmq";
import {
	connectionOptions,
	userActivationRemindersQueueName,
} from "../../helpers/jobsConfiguration";
import { getInactiveUsers } from "../processes/userActivationReminder.process";

export const userActivationRemindersQueue = new Queue(
	userActivationRemindersQueueName,
	{ connection: connectionOptions },
);

export const userActivationReminderProcess = new Worker(
	userActivationRemindersQueueName,
	async (job: Job) => {
		try {
			await getInactiveUsers(job);
		} catch (err) {
			console.log(err);
		}
	},
	{ connection: connectionOptions },
);

// Export trigger for one-off jobs 
export const runActivationRemindersJob = (data?: any) => {
	userActivationRemindersQueue.add("Custom Job", data);
};

userActivationRemindersQueue.on("error", (err) => {
	console.log("User Activation Reminders job quit: ", err);
});
```

I've tried various variations of reusing redis connection, not reusing, setting directly in the queue file, splitting worker and queue into separate files (that worked even worse), and I'm getting to wits end. 

I realize this is an old issue but since it describes the same symptoms I'm finding, I thought it wiser to add on to this one.

**Comment by manast** [comment]
@aarhusgregersen the code above is actually not so helpful. I guess that if this is a connection-related issue the code to reproduce it should not be more than 4 or 5 lines. We also would need more details on how to reproduce the error itself, step by step.

**Comment by manast** [comment]
@aarhusgregersen also, if you haven't already I recommend you to read this: https://docs.bullmq.io/guide/going-to-production#automatic-reconnections

**Comment by aarhusgregersen** [comment]
Thanks @manast , my apologies, I'll see if I can follow up with some better examples. 
I have already been through the going to production checklist several times, and tried so many different combinations of settings and nothing seems to work.

A question though, that I couldn't find anywhere in the documentation: Is bullmq required to run _outside_ the scope of the express server?
I was thinking about this because [node-schedule](https://github.com/node-schedule/node-schedule) says the following in their documentation:
"Note that Node Schedule is designed for in-process scheduling, i.e. scheduled jobs will only fire as long as your script is running,"

**Comment by manast** [comment]
There is no reason why you couldn't run a express server and a BullMQ worker in the same node process if that is what you want to do.

**Comment by Totalus** [comment]
Is it expected that queue commands (ex: `queue.add()`) throw an error if there is a `queue.on('error', ...)` listener defined ? 

### Description

If the two following conditions are met:
- `maxRetriesPerRequest` is not set to `null` in the redis connection options
- Redis connection is lost **after being established at least once**

Then queue commands (ex: `queue.add()`, `queue.getActiveCount()`, ...) will throw an error.

### Reproducing

The following code can be used to reproduce the issue:

```
import { Queue } from "bullmq";

const QUEUE_NAME="myQueue"

const queue = new Queue(QUEUE_NAME, {
    connection: { 
        host: "localhost",
        port: "6379",
        password: '',
        retryStrategy: (times: number) => 200,
        maxRetriesPerRequest: 3,
        offlineQueue: false
    },
});

queue.on('error', error => { console.error(`BullMQ Queue error: ${error}`) });

setInterval(async () => {
    console.log('Doing something on the queue');
    await queue.add(QUEUE_NAME, { Hello: "world" });
}, 5000);
```

1. Ensure the connection to redis is established
2. Kill the connection to redis (ex: shutdown the instance)
3. Shortly after the console prints _"Doing something on the queue"_, `queue.add()` will throw an error: `MaxRetriesPerRequestError: Reached the max retries per request limit (which is 3). Refer to "maxRetriesPerRequest" option for details.`

### Configuration

- Windows 11
- BullMQ 4.11.4
- Node 18.17.1

**Comment by manast** [comment]
@Totalus  This works as design. If you do not want the method to fail use null for the ```maxRetriesPerRequest```. Here some more info: https://docs.bullmq.io/patterns/failing-fast-when-redis-is-down

I am closing this issue as there are not actions for us to take here.

**Comment by AbrahamLopez10** [comment]
@aarhusgregersen Were you able to sort out the issue with getting ETIMEDOUT error in production while BullMQ and Redis appear to be working normally? 

I'm getting the same issue, for reference I'm using AWS MemoryDB to run Redis, BullMQ works normally and jobs are getting processed but I'm getting a `connect ETIMEDOUT` error happening exactly **twice every 12 seconds.** I have already checked job status, Redis metrics, etc. and everything appears normal. I only have one server instance running Bull and I don't have any recurring jobs running on that interval of 12 seconds

@Totalus Does something comes to mind based on your experience on why I could be getting these connect errors so often while everything appears to be working normally?

**Comment by Totalus** [comment]
@AbrahamLopez10 With the number of timeout errors you have, I'm wondering if your server can reach the Redis instance at all. My guess would be that BullMQ is using two connections for whatever operation you ask it to achieve and both can't reach the server so they timeout. That's the only thing that comes to my mind at the moment with the info you shared.

**Comment by aarhusgregersen** [comment]
I was not able to resolve it unfortunately, @AbrahamLopez10 , and instead chose to revert to a much simpler queue using `node-schedule` and our database.

Someone suggested I made an attempt at starting the queue _after_ redis had been initialized, by calling it from within this block:

```
redisClient.on("connect", () => {
	console.log("Connected to redis");
});
```

but I never got around to it, as I had already moved solutions when the suggestion was made.
It seems like it might work and probably I shouldn't have attempted to connect jobs, at least in production, until the redis connection is made.

Hope that helps!

**Comment by AbrahamLopez10** [comment]
@Totalus @aarhusgregersen Thanks for the answers! I'll look into what you both mentioned and will report back if there's any useful findings 🤞

**Comment by manast** [comment]
Just to clarify, it should not be needed to wait for connections to be established with Redis before you can start interacting with the queues, this is handled automatically by IORedis.

**Comment by aarhusgregersen** [comment]
I completely agree @manast , which is I joined in on the discussion of this case, simply because I was seeing unexpected errors. In the end I rolled back to a simpler setup. 
I suspect it could have something to do with our hosting environment, so I have been in the process of trying to replicate on another, more stable, environment (Azure) to see if that has any affect. 
It _shouldn't_ have, but again I also _shouldn't_ be seeing the errors I am, so. 

I'll report back once/if I get around to it, but its currently on the backburner.
